pkg/utils/io/sysconfig.go

```diff
@@ -0,0 +1,139 @@
+package io
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+const defaultSysfsPath = "/sys"
+
+var (
+	SysConfigNotAllowedErr = errors.New("sysConfig key not allowed")
+)
+
+// ParseSysConfigs accepts s in the format 'key1=value1,key2=value2'.
+func ParseSysConfigs(s string, allow func(key string) bool) (map[string]string, error) {
+	s = strings.TrimSpace(s)
+	if s == "" {
+		return nil, nil
+	}
+	result := make(map[string]string)
+	for _, kv := range strings.Split(s, ",") {
+		key, value, found := strings.Cut(kv, "=")
+		if !found || key == "" || value == "" {
+			return nil, errors.New("invalid sysconfig format")
+		}
+		// Prevent users from accessing unexpected files through:
+		// 1. "subsystem" symlinks like /sys/block/vda/subsystem/vdb
+		// 2. relative paths containing ".." like ../../../root/.ssh/id_rsa
+		if strings.Contains(key, "..") || strings.Contains(key, "subsystem") || !allow(key) {
+			return nil, SysConfigNotAllowedErr
+		}
+		result[key] = value
+	}
+	return result, nil
+}
+
+type SysConfigManager struct {
+	sysfsPath    string
+	nonBlockFs   bool
+	major, minor uint32
+}
+
+func NewSysConfigManagerForDisk(device string) (*SysConfigManager, error) {
+	devices := &RealDevTmpFS{}
+	major, minor, err := devices.DevFor(device)
+	if err != nil {
+		return nil, err
+	}
+	return &SysConfigManager{
+		sysfsPath: defaultSysfsPath,
+		major:     major,
+		minor:     minor,
+	}, nil
+}
+
+func NewSysConfigManagerForNFS(mountPath string) (*SysConfigManager, error) {
+	var stat unix.Statx_t
+	err := unix.Statx(0, mountPath, 0, unix.STATX_BASIC_STATS, &stat)
+	if err != nil {
+		return nil, err
+	}
+
+	// check the filesystem is NFS
+	major, minor := stat.Dev_major, stat.Dev_minor
```

Comment:
This should also work for the disk driver. Maybe we don't need a separate `NewSysConfigManagerForDisk`?